### PR TITLE
List bundles

### LINKED ISSUE #254: Make opcap capable of listing operator versions available in a given catalog with their respective channels
## Is your feature request related to a problem? Please describe.
In order to install an operator with a specific version and sequentially test upgrades on that particular operator we need at first to discover what versions are available on a given catalog for that operator. That list will also serve to inform operator developers wanting to troubleshoot their operators and operator bundles.

## Describe the solution you'd like.
A new command and sub command on opcap `bundle list` would take catalogs, channels and also package name if necessary to list all the bundles and consequentially all the versions available for that specific operator. That can be done by using the operator registry api github.com/operator-framework/operator-registry/pkg/client

## Describe alternatives you've considered.
I considered using the OLM clientset api but no information about specific bundles are provided. We can supposedly try the same with just the currentCSV fields on the package/channel combination to create new subscriptions for that startingCSV. That's not simple though. We need to understand all the CSV semantic names to put them in order and try upgrades on that.

## Steps to make this feature available:

- [x] #283
- [x] #284
- [x] #285

### PR DESCRIPTION
**Pull request by acmenezes**
## Add new command and functionality to list bundles with versions from certified-operators repository

Fixes #254  

## Changes (required)

- Add GitCloneBundles
- ADD listBundles function
- ADD list bundles command
- MOVE client creation to separate function
- ADD tests

## Checklist (required)
- [x] I have reviewed and followed the [contribution guidelines](https://github.com/opdev/opcap/blob/main/docs/contribution.md)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation if needed
- [x] I have checked that my changes pass all tests

### CODE REVIEW

File: internal/bundle/bundle.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+			}
+			bundles = append(bundles, bundle)
+		}
+
+	}
+	return bundles, nil
+}
+
+func getStartingCsv(filePath string) (string, error) {
+
```
Comment (bcrochet) [inline internal/bundle/bundle.go:140]: Extra line. Run gofumpt.

File: internal/bundle/bundle.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+	if err = decoder.Decode(&csv); err != nil {
+		return "", err
+	}
+
+	return csv.Name, nil
+}
+
+func getAnnotations(filePath string) (map[string]string, error) {
+
```
Comment (bcrochet) [inline internal/bundle/bundle.go:158]: Extra line. Run gofumpt.

File: internal/bundle/bundle_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
+package bundle
+
+import (
+	. "github.com/onsi/ginkgo/v2/dsl/core"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Package Bundle", func() {
+
```
Comment (bcrochet) [inline internal/bundle/bundle_test.go:9]: Extra line. Run gofumpt.

File: internal/bundle/bundle.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	}
+	err = workTree.Pull(&git.PullOptions{RemoteName: "origin"})
+	if err != nil && err != git.NoErrAlreadyUpToDate {
+		return fmt.Errorf("failed pulling data from repository: %s", err)
+	}
+	return nil
+}
+
+func listBundles(bundlesPath string) ([]bundle, error) {
+
```
Comment (bcrochet) [inline internal/bundle/bundle.go:61]: Extra line. Run gofumpt.

File: internal/bundle/bundle.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	git "gopkg.in/src-d/go-git.v4"
+)
+
+const (
+	URL       = "https://github.com/redhat-openshift-ecosystem/certified-operators.git"
+	outputDir = "internal/bundle/.test/repo/certified-operators"
+)
+
+func List(bundlesPath string) ([]bundle, error) {
+
```
Comment (bcrochet) [inline internal/bundle/bundle.go:21]: Extra line. Run gofumpt.

File: cmd/list_bundles_test.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package cmd
+
+import (
+	. "github.com/onsi/ginkgo/v2/dsl/core"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("List Bundles Cmd", func() {
+
```
Comment (bcrochet) [inline cmd/list_bundles_test.go:9]: Extra line. Run gofumpt.

File: cmd/list_bundles.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+			}
+			w.Flush()
+
+			return nil
+		},
+	}
+
+	flags := cmd.Flags()
+
+	flags.StringVar(&listBundlesFlags.bundlesPath, "bundles-path", "internal/bundle/.test/repo/certified-operators/operators",
```
Comment (bcrochet) [inline cmd/list_bundles.go:43]: Should this be a directory? Or a git repo? Or support both?

File: internal/bundle/bundle.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	operatorv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	git "gopkg.in/src-d/go-git.v4"
+)
+
+const (
+	URL       = "https://github.com/redhat-openshift-ecosystem/certified-operators.git"
+	outputDir = "internal/bundle/.test/repo/certified-operators"
```
Comment (bcrochet) [inline internal/bundle/bundle.go:17]: This should probably just be a temporary dir that gets cleaned up.

**Review by bcrochet — CHANGES_REQUESTED** [review]
A few linter issues and a couple of suggestions/questions.

File: cmd/list_bundles.go (hunk 7 above)
Comment (acmenezes) [reply cmd/list_bundles.go]: Directory. Name changed.

File: internal/bundle/bundle.go (hunk 1 above)
Comment (acmenezes) [reply internal/bundle/bundle.go]: Done.

File: internal/bundle/bundle.go (hunk 2 above)
Comment (acmenezes) [reply internal/bundle/bundle.go]: Done.

File: internal/bundle/bundle.go (hunk 4 above)
Comment (acmenezes) [reply internal/bundle/bundle.go]: Done.

File: cmd/list_bundles_test.go (hunk 6 above)
Comment (acmenezes) [reply cmd/list_bundles_test.go]: Done.

File: cmd/list_bundles.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+			}
+
+			return nil
+		},
+	}
+
+	flags := cmd.Flags()
+
+	flags.StringVar(&listBundlesFlags.bundlesDir, "bundles-path", "internal/bundle/.test/certified-operators",
+		"specifies the source directory with bundles")
```
Comment (bcrochet) [inline cmd/list_bundles.go:37]: I'm not understanding the utility of this flag. Here, it's stated that it is the source directory for bundles. To me, that means that it is expected to already contain the bundles. However, it is not expected to already contain the repo. Is there a reason that the user would need this to be a specific directory? Does it need to persist beyond the execution of the program?

**Review by bcrochet — COMMENTED** [review]
Probably also time to collapse to a single commit.

File: cmd/list_bundles.go (hunk 9 above)
Comment (acmenezes) [reply cmd/list_bundles.go]: You are right here as well. It's changed. Now we have two mutually exclusive flags: --from-dir --from-repo which allows the user to use a dir with bundles or a different repo from certified operators.

File: cmd/list_bundles.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+			}
+			return nil
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&listBundlesFlags.bundlesDir, "from-dir", "",
+		"specifies the source directory with bundles under the operators directory, can't be used with --from-repo")
+	flags.StringVar(&listBundlesFlags.bundlesRepo, "from-repo", "https://github.com/redhat-openshift-ecosystem/certified-operators.git",
+		"Git repository URL from where to download bundles, can't be used with --from-dir")
```
Comment (bcrochet) [inline cmd/list_bundles.go:44]: These can be marked as mutually exclusive and automatically checked.

https://pkg.go.dev/github.com/spf13/cobra#Command.MarkFlagsMutuallyExclusive

File: cmd/list_bundles.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+	return &cmd
+}
+
+func listBundles(ctx context.Context, out io.Writer) error {
+	if listBundlesFlags.bundlesDir != "" {
+
```
Comment (bcrochet) [inline cmd/list_bundles.go:51]: Unnecessary empty line.

File: cmd/list_bundles.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+	bundles, err := bundle.ReadBundlesFromDir(dir)
+	if err != nil {
+		return err
+	}
+	printBundles(bundles, out)
+	return nil
+}
+
+func printBundles(bundles []bundle.Bundle, out io.Writer) {
+
```
Comment (bcrochet) [inline cmd/list_bundles.go:79]: Unnecessary empty line.

File: internal/bundle/bundle.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+func ReadBundlesFromDir(bundlesDir string) ([]Bundle, error) {
+	bundles := []Bundle{}
+
+	operators, err := os.ReadDir(filepath.Join(bundlesDir, "operators"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract operators from repo: %s", err)
+	}
+
+	for _, operator := range operators {
+
```
Comment (bcrochet) [inline internal/bundle/bundle.go:48]: Unnecessary empty line.

File: internal/bundle/bundle.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		path := filepath.Join(bundlesDir, "operators", operator.Name())
+
+		versions, err := os.ReadDir(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract versions for operator: %s", err)
+		}
+
+		// looking into each version folder
+		for _, version := range versions {
+
```
Comment (bcrochet) [inline internal/bundle/bundle.go:61]: Unnecessary empty line.

File: internal/bundle/bundle.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+			bundle.Version = version.Name()
+			versionDir := filepath.Join(path, version.Name())
+
+			data, err := os.ReadDir(versionDir)
+			if err != nil {
+				return nil, fmt.Errorf("failed to extract bundle for version: %s", err)
+			}
+
+			for _, d := range data {
+
```
Comment (bcrochet) [inline internal/bundle/bundle.go:75]: Unnecessary empty line.

File: internal/bundle/bundle.go [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+				// checking the manifests folder for the csv file
+				if d.IsDir() && d.Name() == "manifests" {
+
```
Comment (bcrochet) [inline internal/bundle/bundle.go:78]: Unnecessary empty line.

**Review by bcrochet — CHANGES_REQUESTED** [review]
Some more gofumpt errors.

File: cmd/list_bundles.go (hunk 11 above)
Comment (acmenezes) [reply cmd/list_bundles.go]: gofumpt didn't complain.

File: cmd/list_bundles.go (hunk 12 above)
Comment (acmenezes) [reply cmd/list_bundles.go]: gofumpt didn't complain.

File: internal/bundle/bundle.go (hunk 13 above)
Comment (acmenezes) [reply internal/bundle/bundle.go:48]: gofumpt didn't complain.

File: internal/bundle/bundle.go (hunk 14 above)
Comment (acmenezes) [reply internal/bundle/bundle.go:61]: gofumpt didn't complain.

File: internal/bundle/bundle.go (hunk 15 above)
Comment (acmenezes) [reply internal/bundle/bundle.go:75]: gofumpt didn't complain.

File: internal/bundle/bundle.go (hunk 5 above)
Comment (acmenezes) [reply internal/bundle/bundle.go]: This function is no longer there.

File: internal/bundle/bundle_test.go (hunk 3 above)
Comment (acmenezes) [reply internal/bundle/bundle_test.go]: I guess this one is also changed.

**Review by bcrochet — APPROVED** [review]

**Pull request merged**